lib/kernel/src/pg.erl

```diff
@@ -117,33 +117,45 @@ start_link(Scope) when is_atom(Scope) ->
 
 %%--------------------------------------------------------------------
 %% @doc
-%% Joins a single process
+%% Joins a single or a list of processes.
 %% Group is created automatically.
-%% Process must be local to this node.
+%% Processes must be local to this node.
 -spec join(Group :: group(), PidOrPids :: pid() | [pid()]) -> ok.
 join(Group, PidOrPids) ->
     join(?DEFAULT_SCOPE, Group, PidOrPids).
 
 -spec join(Scope :: atom(), Group :: group(), PidOrPids :: pid() | [pid()]) -> ok.
-join(Scope, Group, PidOrPids) ->
-    Node = node(),
-    is_list(PidOrPids) andalso [error({nolocal, Pid}) || Pid <- PidOrPids, node(Pid) =/= Node orelse not is_pid(Pid)],
+join(Scope, Group, PidOrPids) when is_pid(PidOrPids); is_list(PidOrPids) ->
+    ok = ensure_local(PidOrPids),
     gen_server:call(Scope, {join_local, Group, PidOrPids}, infinity).
 
 %%--------------------------------------------------------------------
 %% @doc
-%% Single process leaving the group.
-%% Process must be local to this node.
+%% Single or list of processes leaving the group.
+%% Processes must be local to this node.
 -spec leave(Group :: group(), PidOrPids :: pid() | [pid()]) -> ok.
 leave(Group, PidOrPids) ->
     leave(?DEFAULT_SCOPE, Group, PidOrPids).
 
--spec leave(Scope :: atom(), Group :: group(), Pid :: pid() | [pid()]) -> ok | not_joined.
-leave(Scope, Group, PidOrPids) ->
-    Node = node(),
-    is_list(PidOrPids) andalso [error({nolocal, Pid}) || Pid <- PidOrPids, node(Pid) =/= Node orelse not is_pid(Pid)],
```

Comment:
Could this be fixed by swapping the order of `not is_pid(Pid)` and `node(Pid) =/= Node`? This should ensure it properly returns in case the argument is not a pid instead of silently failing.

Reply:
That does not fix the issue as it is an `orelse` and even if `not is_pid(Pid)` is `true`, any `Pid` input to `node/1` there will cause the entire filter to become `false`. It could probably be massaged by different grouping and `not` placement, but it is a bit unwieldy in my opinion with the special casing where the exceptions are eaten and all. The case of where it is a pid could be made proper by just rearranging, but the good case already works, so it'd achieve little. Even then, the case of a single pid wouldn't be checked as line 144 is only concerned if it is a list.